packages/@react-stately/toast/src/useToastState.ts

```diff
@@ -10,57 +10,233 @@
+  remove(key: string): void,
+  /** Pauses the timers for all visible toasts. */
+  pauseAll(): void,
+  /** Resumes the timers for all visible toasts. */
+  resumeAll(): void,
+  /** The visible toasts. */
+  visibleToasts: QueuedToast<T>[]
+}
+
+/**
+ * Provides state management for a toast queue. Toasts are transient notifications
+ * of actions, errors, or other events in an application.
+ */
+export function useToastState<T>(props: ToastStateProps = {}): ToastState<T> {
+  let {maxVisibleToasts = 1, hasExitAnimation = false} = props;
+  let queue = useMemo(() => new ToastQueue<T>({maxVisibleToasts, hasExitAnimation}), [maxVisibleToasts, hasExitAnimation]);
+  return useToastQueue(queue);
+}
 
-export function useToastState(props?: ToastStateProps): ToastState {
-  const [toasts, setToasts] = useState(props && props.value || []);
-  const toastsRef = useRef(toasts);
-  toastsRef.current = toasts;
+/**
+ * Subscribes to a provided toast queue and provides methods to update it.
+ */
+export function useToastQueue<T>(queue: ToastQueue<T>): ToastState<T> {
+  let subscribe = useCallback(fn => queue.subscribe(fn), [queue]);
+  let getSnapshot = useCallback(() => queue.visibleToasts, [queue]);
+  let visibleToasts = useSyncExternalStore(subscribe, getSnapshot);
 
-  const onAdd = (content: ReactNode, options: ToastProps) => {
-    let tempToasts = [...toasts];
-    let timer;
+  return {
+    visibleToasts,
+    add: (content, options) => queue.add(content, options),
+    close: key => queue.close(key),
+    remove: key => queue.remove(key),
+    pauseAll: () => queue.pauseAll(),
+    resumeAll: () => queue.resumeAll()
+  };
+}
 
-    // set timer to remove toasts
-    if (!(options.actionLabel || options.timeout === 0)) {
-      if (options.timeout < 0) {
-        options.timeout = TOAST_TIMEOUT;
+/**
+ * A ToastQueue is a priority queue of toasts.
+ */
+export class ToastQueue<T> {
+  private queue: QueuedToast<T>[] = [];
+  private subscriptions: Set<() => void> = new Set();
+  private maxVisibleToasts: number;
+  private hasExitAnimation: boolean;
+  /** The currently visible toasts. */
+  visibleToasts: QueuedToast<T>[] = [];
+
+  constructor(options?: ToastStateProps) {
+    this.maxVisibleToasts = options?.maxVisibleToasts ?? 1;
+    this.hasExitAnimation = options?.hasExitAnimation ?? false;
+  }
+
+  /** Subscribes to updates to the visible toasts. */
+  subscribe(fn: () => void) {
+    this.subscriptions.add(fn);
+    return () => this.subscriptions.delete(fn);
+  }
+
+  /** Adds a new toast to the queue. */
+  add(content: T, options: ToastOptions = {}) {
+    let toastKey = Math.random().toString(36);
+    let toast: QueuedToast<T> = {
+      ...options,
+      content,
+      key: toastKey,
+      timer: options.timeout ? new Timer(() => this.close(toastKey), options.timeout) : null
+    };
+
+    let low = 0;
+    let high = this.queue.length;
+    while (low < high) {
+      let mid = Math.floor((low + high) / 2);
```

Comment:
can't this possibly insert same priority toasts ahead of where they should be?
would it make more sense to track how many of each priority there are? or keep multiple different queues of priority?
That could simplify the +4 you give to priority of actionable toasts

Reply:
No, same priority toasts are placed behind due to `>` check below.